### ISSUE TITLE
Make CSS more maintainable

### ISSUE THREAD

**Issue by Garbee** [devtools-html/debugger.html#1869]
The CSS as we are doing things now has three primary problems for long-term maintenance.

1. The bidirectional postcss filter is wreaking havoc on our specificity. Which means we need to have a hawk eye on changes or UX problems can easily slip in.
2. We can end up repeating some things often. Which, if ever rewritten means a lot of little line changes to keep up with.
3. Selectors are all over the place and are feeble to any little DOM changes in the future.

A solution that will start to help with the first two issues is, utilizing Sass (via `node-sass`) under the hood for our styles. The output of that can still be passed through PostCSS for things like vendor prefixes and other newer web-tech goodies if needed.

Beyond converting to Sass we can begin to utilize BEM style selectors to keep the selectors as targeted as possible. In some cases, we simply can't achieve this nicely (like targeting icons to get them restyled properly) since they'ed inherently break BEM. But for the vast majority of our components, adhering to BEM will make our code easier to read and maintain.

An example conversion is `CommandBar.css`:

<details>
<summary>Sass and no bidirectional</summary>

```scss
.command-bar {
  height: 30px;
  border-bottom: 1px solid var(--theme-splitter-color);
  display: flex;

  & > button {
    appearance: none;
    background: transparent;
    border: none;
    cursor: pointer;
    display: inline-block;
    text-align: center;
    transition: all 0.25s ease;
    padding: 8px 5px;
    position: relative;
    fill: currentColor;
    // Instead of `margin-inline-end`
    margin: 0 0.7em;

    &:disabled {
      opacity: 0.3;
      cursor: default;
    }

    :root.theme-dark & {
      color: var(--theme-body-color);
    }

    & > i {
      height: 100%;
      width: 100%;
      display: block;

      & > svg {
        width: 16px;
        height: 16px;
      }
    }

    &.pause-exceptions.uncaught {
      color: var(--theme-highlight-purple);
    }

    &.pause-exceptions.all {
      color: var(--theme-highlight-blue);
    }
  }

  html:not([dir="rtl"]) &.pause-exceptions {
    margin-left: .5em;
  }

  html[dir="rtl"] &.pause-exceptions {
    margin-right: .5em;
  }
}
```

</details>

Going to Sass with no bidirectional means we have clear targets that don't get made into something more specific than we say later on (and silently.) This also makes modifying things faster since if the DOM structure changes, we touch far fewer lines to get the styles updated.

<details>
<summary>Sass with BEM classes</summary>

```scss
.command-bar {
  height: 30px;
  border-bottom: 1px solid var(--theme-splitter-color);
  display: flex;

  & > .command-bar__action {
    appearance: none;
    background: transparent;
    border: none;
    cursor: pointer;
    display: inline-block;
    text-align: center;
    transition: all 0.25s ease;
    padding: 8px 5px;
    position: relative;
    fill: currentColor;
    // Instead of `margin-inline-end`
    margin: 0 0.7em;

    &:disabled {
      opacity: 0.3;
      cursor: default;
    }

    :root.theme-dark & {
      color: var(--theme-body-color);
    }

    & > i {
      height: 100%;
      width: 100%;
      display: block;

      & > svg {
        width: 16px;
        height: 16px;
      }
    }

    &--pause-on-uncaught {
      color: var(--theme-highlight-purple);
    }

    &--pause-on-all {
      color: var(--theme-highlight-blue);
    }
  }

  /** No need to convert this to BEM, difficult targets to get right anyways */
  html:not([dir="rtl"]) &.pause-exceptions {
    margin-left: .5em;
  }

  html[dir="rtl"] &.pause-exceptions {
    margin-right: .5em;
  }
}
```

</details>

Adding BEM in on top (where applicable) means we then gain direct targets that are not caught up in the DOM structure. So if we do change elements or nesting, we just update the rules as-needed to modify the appearance to what is now needed without changing the selectors to get to it.

Are there any other approaches we could use to make the styles more maintainable? I do know PostCSS has a nesting plugin, so using that could be an option over integrating Sass. However the syntax can be quirky if we use the syntax Tab is writing a proposal for the CSS WG about. In my experience with Google's Material Components for Web when we were starting the new architecture for that, Sass helps making small maintainable components much easier than what PostCSS handles.

**Comment by irfanhudda** [comment]
+1 for more maintainable CSS. Now that we are refactoring react components, we should also refactor CSS.

It might be good to also have CSS that is react component specific.(i.e. doesn't affect whole html page) It would help refactoring and maintaining in future.

**Comment by Garbee** [comment]
> CSS that is react component specific.(i.e. doesn't affect whole html page)

This is what BEM does for us. It helps enforce isolated pieces. Currently, only the text direction stuff needs to be handled from the `html` scope check. However, that is not 100% accurate (as the mode *could* change internally.) Which is where the `dir` [psuedo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/:dir) is very handy. However, only support in Firefox currently so that is a harder-to-use item since we are looking to have the UX run in other engines if manually built. That is something we can have as a `TODO` to refactor later when support for that gets wider.

**Comment by irfanhudda** [comment]
I'm not very familiar with BEM. But from what I read it might help solve our problem.

If we use sass we should keep selector nesting to minimum. After nesting 2 to 3 times, it becomes hard to follow.

**Comment by Garbee** [comment]
Yup, nesting can be handled through stylelint rules. The reason the commandbar had 4th level nesting is I just didn't care to trim it out for a demo. :wink: Also if we look at some of the places where the styles get too deeply nested and such, it shows we have a thing to abstract into another component to handle something.

**Comment by jasonLaster** [comment]
I'm going to close this issue for the time being as we're already introducing a lot of change in the Debugger. 

We're developing an RFC process for discussing project level changes. This will be a good candidate at the right time.

I'm also adding this issue as a note in our [tidy up] project board.

[tidy up]: https://github.com/devtools-html/debugger.html/projects/4
